File: open_ai/1.cd_directory_v3.py
import queue

class Solution:
    def simplifyPath(self, path):
        # tokens = path.split('/')
        tokens = filter(lambda x : x != '.' and x != '', path.split('/'))
        result = []
        for token in tokens:
            if token == "..":
                if len(result) > 0: result.pop()
                else: return 'Null'
            else:
                result.append(token)

        return '/' + '/'.join(result)

    
    def cd(self, current_dir, new_dir):
        if self.simplifyPath(current_dir) == 'Null' or self.simplifyPath(new_dir) == 'Null': 
            return 'Null'
        return self.simplifyPath(current_dir + '/' + new_dir)

    
    def hasCycle(self, soft_links):
        """ O(n) Kahn algorithm: Topological sorting. """
        inDegree = {}
        outDegree = {}
        links = set()
        for key, value in soft_links.items():
            if not key in outDegree:
                outDegree[key] = set()
            outDegree[key].add(value)
            inDegree[value] = inDegree.get(value, 0) + 1
        visited = set()
        stack = [k for k, v in inDegree.items() if v == 0]
        while stack:
            node = stack.pop()
            visited.add(node)
            for neighbor in outDegree[node]:
                inDegree[neighbor] -= 1
                if inDegree[neighbor] == 0:
                    stack.append(neighbor)
        return len(visited) != len(soft_links)

    
    def cdWithSoftLinks(self, current_dir, new_dir, soft_links):
        # print(soft_links)
        if self.hasCycle(soft_links):
            return 'Cycle Detected'
        
        result = self.simplifyPath(current_dir + "/" + new_dir)
        # print("result: ", result)
        if result in soft_links:
            # print('resul tin soft_link')
            while result in soft_links:
                result = soft_links[result]
            return result
        
        while current_dir in soft_links:
            current_dir = soft_links[current_dir]
        while new_dir in soft_links:
            new_dir = soft_links[new_dir]
        result = self.simplifyPath(current_dir + "/" + new_dir)
        while result in soft_links:
            result = soft_links[result]
        return result


    def hasCycle2(self, soft_links):
        inDegree = {}
        outDegree = {}
        links = set()
        for key, value in soft_links.items():
            if not key in outDegree:
                outDegree[key] = set()
            outDegree[key].add(value)
            if not value in inDegree:
                inDegree[value] = set()
            inDegree[value].add(key)

            links.add(key)
            links.add(value)
        
        q = queue.Queue()
        for link in links:
            if not link in inDegree or len(inDegree[link]) == 0:
                q.put(link)
        
        visited = set()
        
        while not q.empty():
            link = q.get()
            visited.add(link)
            
            if link in outDegree:
                for nextLink in outDegree[link]:
                    inDegree[nextLink].remove(link)
                    if not nextLink in inDegree or len(inDegree[nextLink]) == 0:
                        q.put(nextLink)

        # print("len(links) = ", len(links), ", len(visited) = ", len(visited))
        return len(links) > len(visited)


    def hasCycle3(self, soft_links):
        inDegree = {}
        outDegree = {}
        links = set()
        for key, value in soft_links.items():
            if not key in outDegree:
                outDegree[key] = set()
            outDegree[key].add(value)
            if not value in inDegree:
                inDegree[value] = set()
            inDegree[value].add(key)

            links.add(key)
            links.add(value)
        
        q = queue.Queue()
        for link in links:
            if not link in inDegree or len(inDegree[link]) == 0:
                # print('q.put: ' + link)
                q.put(link)
        
        while not q.empty():
            link = q.get()
            # print('remove: link = ' + link)
            links.remove(link)

            if link in outDegree:
                for nextLink in outDegree[link]:
                    inDegree[nextLink].remove(link)
                    if not nextLink in inDegree or len(inDegree[nextLink]) == 0:
                        q.put(nextLink)

        # print("links: ", links)
        return len(links) > 0
    

class SolutionTest:
    def testSimplifyPath(self, path, expect):
        so = Solution()
        result = so.simplifyPath("/abc/../c/././d")
        if result == expect:
            print('Pass: result = ' + result)
        else:
            print('Fail: result = ' + result + ', expect = ' + expect)

    def testCd(self, current_dir, new_dir, expect):
        so = Solution()
        result = so.cd(current_dir, new_dir)
        if result == expect:
            print('Pass: result = ' + result)
        else:
            print('Fail: result = ' + result + ', expect = ' + expect)
    
    def testCdWithSoftLinks(self, current_dir, new_dir, soft_links, expect):
        so = Solution()
        result = so.cdWithSoftLinks(current_dir, new_dir, soft_links)
        if result == expect:
            print('Pass: result = ' + result)
        else:
            print('Fail: result = ' + result + ', expect = ' + expect)
    
    def runTests(self):
        self.testSimplifyPath("/abc/../c/././d", "/c/d")
        self.testCd("/a", "/b", "/a/b")
        self.testCd("/foo/bar", "baz", "/foo/bar/baz")
        self.testCd("/foo/../", "./baz", "/baz")
        self.testCd("/", "foo/bar/../../baz", "/baz")
        self.testCd("/", "..", "Null")
        self.testCd("/a/../..", "b", "Null")
        self.testCd("/a/", "/b/../..", "Null")

        soft_links = {"/foo/bar": "/abc"}
        self.testCdWithSoftLinks("/foo/bar", "baz", soft_links, "/abc/baz")

        soft_links = {"/foo/bar": "/abc", "/abc": "/bcd", "/bcd/baz": "/xyz"}
        self.testCdWithSoftLinks("/foo/bar", "baz", soft_links, "/xyz")

        # cd(/foo/bar, baz, {/foo/bar: /abc, /foo/bar/baz: xyz}) = xyz
        soft_links = {"/foo/bar": "/abc", "/foo/bar/baz": "xyz"}
        self.testCdWithSoftLinks("/foo/bar", "baz", soft_links, "xyz")

        soft_links = {"/foo/bar": "/abc", "A": "B", "/abc": "/foo/bar", "B": "A"}
        self.testCdWithSoftLinks("/foo/bar", "baz", soft_links, "Cycle Detected")

        soft_links = {"/foo/bar": "/abc", "/abc": "/d", "/d": "/foo/bar", "B": "A"}
        self.testCdWithSoftLinks("/foo/bar", "baz", soft_links, "Cycle Detected")

        soft_links = {"A": "B", "B": "A"}
        self.testCdWithSoftLinks("/foo/bar", "baz", soft_links, "Cycle Detected")

        soft_links = {"A": "A", "B": "C"}
        self.testCdWithSoftLinks("/foo/bar", "baz", soft_links, "Cycle Detected")

if __name__ == "__main__":
    test = SolutionTest()
    test.runTests()

File: open_ai/1.cd_directory_v3.cpp
// 5:26 - 6:06
// 6:40 - 6:52
#include <iostream>
#include <sstream>
#include <vector>
#include <set>

using namespace std;

class Solution {
public:
    string simplifyPath(string path) {
        vector<string> tokens;
        string token;
        stringstream ss(path);
        while (getline(ss, token, '/')) {
            if (token == "" || token == ".") {
                // do nothing
            } else if (token == "..") {
                if (! tokens.empty()) tokens.pop_back();
                else return "Null";
            } else {
                tokens.push_back(token);
            }
        }

        string result = "/";
        for (string token : tokens) {
            if (result != "/") result += "/";
            result += token;
        }
        return result;
    }

    string cd(string current_dir, string new_dir, unordered_map<string, string>& soft_links) {
        if (hasCycle(soft_links)) {
            return "Cycle Detected";
        }

        string result = simplifyPath(current_dir + "/" + new_dir);
        if (soft_links.count(result)) { // has long match.
            while (soft_links.count(result)) {
                result = soft_links[result];
            }
            return result;
        }

        // no long match
        while (soft_links.count(current_dir)) {
            current_dir = soft_links[current_dir];
        }
        while (soft_links.count(new_dir)) {
            new_dir = soft_links[new_dir];
        }
        result = simplifyPath(current_dir + "/" + new_dir);
        while (soft_links.count(result)) {
            result = soft_links[result];
        }
        return result;
    }

    string cd(string current_dir, string new_dir) {
        if (simplifyPath(current_dir) == "Null" || simplifyPath(new_dir) == "Null") {
            return "Null";
        }
        return simplifyPath(current_dir + "/" + new_dir);
    }

private:
    bool hasCycle(unordered_map<string, string>& soft_links) {
        unordered_map<string, set<string>> inDegree, outDegree;
        set<string> links;

        for (auto item : soft_links) {
            inDegree[item.second].insert(item.first);
            outDegree[item.first].insert(item.second);

            links.insert(item.first);
            links.insert(item.second);
        }

        queue<string> q;
        for (string link : links) {
            if (inDegree[link].empty()) {
                q.push(link);
            }
        }

        while (! q.empty()) {
            string link = q.front();
            q.pop();

            links.erase(link);

            for (string nextLink : outDegree[link]) {
                inDegree[nextLink].erase(link);

                if (inDegree[nextLink].empty()) {
                    q.push(nextLink);
                }
            }
        }

        return ! links.empty();
    }
};

class SolutionTest {
private:
    void testCdWithSoftLinks(string current_dir, string new_dir, unordered_map<string, string>& soft_links, string expect) {
        Solution so;
        string result = so.cd(current_dir, new_dir, soft_links);
        
        if (result == expect) {
            cout << "Pass: result = " << result  << endl;
        } else {
            cout << "Fail: result = " << result << ", expect = " << expect << endl;
        }
    }

    void testCd(string current_dir, string new_dir, string expect) {
        Solution so;
        string result = so.cd(current_dir, new_dir);
        
        if (result == expect) {
            cout << "Pass: result = " << result  << endl;
        } else {
            cout << "Fail: result = " << result << ", expect = " << expect << endl;
        }
    }

    void testSimplifyPath(string path, string expect) {
        Solution so;
        string result = so.simplifyPath(path);

        if (result == expect) {
            cout << "Pass: input: " << path << ", result = " << result  << endl;
        } else {
            cout << "Fail: input: " << path << ", result = " << result << ", expect = " << expect << endl;
        }
    }

public:
    void run_tests() {
        testSimplifyPath("/abc/../c/././d", "/c/d");
        testSimplifyPath("/abc/./c/././d", "/abc/c/d");
        testSimplifyPath("/abc/../", "/");

        testCd("/a", "/b", "/a/b");

        // cd(/foo/bar, baz) = /foo/bar/baz
        testCd("/foo/bar", "baz", "/foo/bar/baz");

        // cd(/foo/../, ./baz) = /baz
        testCd("/foo/../", "./baz", "/baz");

        // cd(/, foo/bar/../baz) = /baz
        testCd("/", "foo/bar/../../baz", "/baz");

        // cd(/, ..) = Null
        testCd("/", "..", "Null");

        testCd("/a/", "/b/../..", "Null");

        unordered_map<string, string> soft_links = {{"/foo/bar", "/abc"}};
        testCdWithSoftLinks("/foo/bar", "baz", soft_links, "/abc/baz");

        // cd(/foo/bar, baz, {/foo/bar: /abc, /abc: /bcd, /bcd/baz: /xyz}) = /xyz
        soft_links = {{"/foo/bar", "/abc"}, {"/abc", "/bcd"}, {"/bcd/baz", "/xyz"}};
        testCdWithSoftLinks("/foo/bar", "baz", soft_links, "/xyz");

        // cd(/foo/bar, baz, {/foo/bar: /abc, /foo/bar/baz: xyz}) = xyz
        soft_links = {{"/foo/bar", "/abc"}, {"/foo/bar/baz", "xyz"}};
        testCdWithSoftLinks("/foo/bar", "baz", soft_links, "xyz");

        soft_links = {{"/foo/bar", "/abc"}, {"A", "B"}, {"/abc", "/foo/bar"}, {"B", "A"}};
        testCdWithSoftLinks("/foo/bar", "baz", soft_links, "Cycle Detected");

        soft_links = {{"/foo/bar", "/abc"}, {"/abc", "/d"}, {"/d", "/foo/bar"}, {"B", "A"}};
        testCdWithSoftLinks("/foo/bar", "baz", soft_links, "Cycle Detected");

        soft_links = {{"A", "B"}, {"B", "A"}};
        testCdWithSoftLinks("/foo/bar", "baz", soft_links, "Cycle Detected");

        soft_links = {{"A", "A"}, {"B", "C"}};
        testCdWithSoftLinks("/foo/bar", "baz", soft_links, "Cycle Detected");
    }
};

int main() {
    SolutionTest soTest;
    soTest.run_tests();
    return 0;
}
